File: rplugin/python3/acid/handlers/goto.py
from acid.handlers import BaseHandler
from acid.nvim import find_file_in_path
from acid.nvim.log import log_info, error, warning


class Handler(BaseHandler):

    name = "Goto"
    priority = 0

    def on_handle(self, msg, *_):
        if 'file' in msg:
            f = find_file_in_path(self.nvim, msg)

            if f is None:
                warning(self.nvim, "File not found")
                return

            c = msg.get('column', 1)
            l = msg.get('line', 1)

            current_scrolloff = self.nvim.options['scrolloff']
            self.nvim.options['scrolloff'] = 999

            try:
                if self.nvim.funcs.expand('%').endswith(f):
                    self.nvim.funcs.cursor(l, c)
                else:
                    self.nvim.command("edit +{} {}".format(l, f))
            except Exception as e:
                error(self.nvim, "Error while navigating: {}".format(str(e)))
            finally:
                self.nvim.options['scrolloff'] = current_scrolloff

        elif 'no-info' in msg['status']:
            warning(self.nvim, 'No information found for symbol')

File: lua/acid/middlewares/go_to.lua
-- luacheck: globals vim
local nvim = vim.api
local log = require("acid.log")
local utils = require("acid.utils")
local go_to = {}

go_to.config = {
  open = "edit"
}

go_to.middleware = function(config)
  return function(middleware)
    return function(data)

      if utils.find(data.status, "no-info") then
        log.msg("No information found")
        return
      end

      local args = {data.file}
      if data.resource ~= nil then
        table.insert(args, data.resource)
      end

      local fpath = nvim.nvim_call_function("AcidFindFileInPath", args)

      if fpath == nil then
        log.msg("File not found (" .. data.file .. ").")
        return
      end

      local col = math.floor(data.column or 1)
      local ln = math.floor(data.line or 1)

      local cur_scroll = nvim.nvim_get_option("scrolloff")
      nvim.nvim_set_option("scrolloff", 999)

      if utils.ends_with(nvim.nvim_call_function("expand", {"%"}), fpath) then
        nvim.nvim_call_function("cursor", {ln, col})
      else
        nvim.nvim_command(config.open .. " +" .. ln .. " " .. fpath)
      end

      nvim.nvim_set_option("scrolloff", cur_scroll)

      return middleware(data)
    end
  end
end

return go_to
